contracts/oracles/CurveOracle.sol

```diff
@@ -0,0 +1,48 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.11;
+pragma abicoder v1;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "../interfaces/IOracle.sol";
+import "../libraries/Sqrt.sol";
+
+import "../interfaces/ICurveRegistry.sol";
+import "../interfaces/ICurveProvider.sol";
+import "../interfaces/ICurveSwap.sol";
+
+contract CurveOracle is IOracle {
+    using Sqrt for uint256;
+
+    ICurveProvider public immutable addressProvider;
+    IERC20 private constant _NONE = IERC20(0xFFfFfFffFFfffFFfFFfFFFFFffFFFffffFfFFFfF);
+
+    constructor(ICurveProvider _addressProvider) {
+        addressProvider = _addressProvider;
+    }
+
+    function getRate(IERC20 _srcToken, IERC20 _dstToken, IERC20 connector) external view override returns (uint256 rate, uint256 weight) {
+
+        require(connector == _NONE, "connector should be NONE");
+
+        address registryAddress = addressProvider.get_address(0);
+        ICurveRegistry registry = ICurveRegistry(registryAddress);
+
+        address pool;
+        pool = registry.find_pool_for_coins(address(_srcToken), address(_dstToken));
```

Comment:
`find_pool_for_coins` supports paging in case when there is more than one pool. So here it should iterate until returned pool is not zero

Reply:
Where do you read this is the case? In the documentation this is not mentioned. It will instead return ZERO_ADDRESS if the swap is not available. (https://curve.readthedocs.io/registry-registry.html#Registry.find_pool_for_coins)

Reply:
Exactly in the link you provided. It says:
```You can optionally include i to get the n-th pool, when multiple pools exist for the given pairing.
Returns 0x00 when swaps are not possible for the pair or i exceeds the number of available pools.```

Reply:
Are we talking about the same thing? It says it will return 0x00 if swaps are **not** possible. If there exist multiple pools it will just return the first one. no?

Reply:
I think so. Here is the screenshot of the text I cited:
<img width="730" alt="Screenshot 2022-01-27 at 18 47 45" src="https://user-images.githubusercontent.com/735355/151382349-3f8c09ea-3238-4e48-8554-401a69740491.png">

You should iterate over all the supported pools until the function returns `address(0)`. And after that average the result of all the supported pools.

Reply:
You can check the implementation for Kyber DMM which also supports multiple pools per asset:

https://github.com/1inch/offchain-oracle/blob/5a991c28584a4cf3ca2eff0d1568ea2c391cc311/contracts/oracles/KyberDmmOracle.sol

Reply:
Sure I could do that. But this would be very gas expensive since many assets can be traded on many pools - it would be the same gas expeniture as calling `get_best_rate` which do iterate over all pools to find the best rate. Is it what you'd like to have done? Essentially a `get_best_rate` but for finding the average rate instead?